mlrun/runtimes/api_gateway.py

```diff
@@ -0,0 +1,170 @@
+            self._auth = f"Basic {token}"
+
+    def requires_auth(self):
+        return self._auth is not None
+
+    def _generate_invoke_url(self):
+        nuclio_hostname = urllib.parse.urlparse(self._nuclio_dashboard_url).netloc
+
+        # Remove the 'nuclio' prefix from the hostname
+        # For example, from `nuclio.default-tenant.app.dev62.lab.iguazeng.com`,
+        # it becomes `default-tenant.app.dev62.lab.iguazeng.com`
+        common_hostname = nuclio_hostname[nuclio_hostname.find(".") + 1 :]
+
+        # Generate a unique invoke URL which contains the API gateway name and project name
+        return urllib.parse.urljoin(
+            f"{self.name}-{self.project}.{common_hostname}", self.path
+        )
+
+    @classmethod
+    def from_values(
+        cls,
+        project,
+        name: str,
+        host: str,
+        path: str,
+        description: str,
+        functions: list[str],
+        username: Union[None, str],
+        password: Union[None, str],
+        canary: Union[list[int], None],
+    ) -> "APIGateway":
+        if not name:
+            raise mlrun.errors.MLRunInvalidArgumentError(
+                "API Gateway name cannot be empty"
+            )
+
+        if canary:
+            if len(functions) != len(canary):
+                raise mlrun.errors.MLRunInvalidArgumentError(
+                    "Lengths of function and canary lists do not match"
+                )
+            if sum(canary) != 100:
+                raise mlrun.errors.MLRunInvalidArgumentError(
+                    "The sum of canary function percents should be equal to 100"
+                )
+        return cls(
+            project=project,
+            name=name,
+            host=host,
+            path=path,
+            description=description,
+            functions=functions,
+            username=username,
+            password=password,
+            canary=canary,
+        )
+
+    @classmethod
+    def from_dict(
+        cls,
+        dict_values,
+    ):
+        project = (
+            dict_values.get("metadata", {})
+            .get("labels", {})
+            .get("nuclio.io/project-name")
+        )
+        spec = dict_values.get("spec", {})
+        authentication_mode = spec.get(
+            "authenticationMode", NO_AUTH_NUCLIO_API_GATEWAY_AUTH_MODE
+        )
+        authentication = spec.get("authentication", {}).get(authentication_mode, {})
+        upstreams = spec.get("upstreams", [])
+        functions, canary = cls._process_canary(upstreams)
+        if not functions:
+            return None
+        return cls(
+            project=project,
+            name=spec.get("name"),
+            host=spec.get("host"),
+            path=spec.get("path", ""),
+            description=spec.get("description", ""),
+            functions=functions,
+            username=authentication.get("username"),
+            password=authentication.get("password"),
```

Comment:
Do we expect to get the username/password from the BE response when listing GWs (this is where `from_dict` is used currently)?

Reply:
@theSaarco yes, username/password is returned within BE response. As an example, test `test_list_api_gateways` contains real response tested on my system.

Reply:
This looks like a bad practice that goes all the way to nuclio - @liranbg , @TomerShor - are there any plans to remove credentials data from gw structured returned in nuclio APIs?

Reply:
@theSaarco @TomerShor I'll get rid of all the credentials stored in the APIGateways object and the user will have to pass authentication information when calling the function.